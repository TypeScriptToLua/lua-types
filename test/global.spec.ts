import { describeForEachLuaTarget, tstl } from './test-utils';

describeForEachLuaTarget('global', (target) => {
    test('assert', () => {
        const lua = tstl(
            target,
            `
            assert({ bla: "not false"});
        `
        );

        expect(lua).toMatchSnapshot();
    });

    test('assert with return', () => {
        const lua = tstl(
            target,
            `
            const v = assert({ bla: "not false"});
            const bla = v.bla;
        `
        );

        expect(lua).toMatchSnapshot();
    });

    test('assert with multi-return', () => {
        const lua = tstl(
            target,
            `
            const [v, a, b] = assert({ bla: "not false"}, { foo: "FOO" }, { bar: "BAR" });
            const bla = v.bla;
            const foo = a.foo;
            const bar = b.bar;
        `
        );

        expect(lua).toMatchSnapshot();
    });

    test('assert', () => {
        const lua = tstl(
            target,
            `
            assert(false, "assert message");
        `
        );

        expect(lua).toMatchSnapshot();
    });

    test('getmetatable', () => {
        const lua = tstl(
            target,
            `
            const metatable = getmetatable({});
            const add = metatable!.__add;
        `
        );

        expect(lua).toMatchSnapshot();
    });

    test('getmetatable on string', () => {
        const lua = tstl(
            target,
            `
            const metatable = getmetatable("foo");
            const index = metatable!.__index;
        `
        );

        expect(lua).toMatchSnapshot();
    });

    test('ipairs', () => {
        const lua = tstl(
            target,
            `
            for (const [i, v] of ipairs([1, 2, 3])) {
                print(i, v);
            }
        `
        );

        expect(lua).toMatchSnapshot();
    });

    test('next', () => {
        const lua = tstl(
            target,
            `
            const [nextKey, nextValue] = next({});
        `
        );

        expect(lua).toMatchSnapshot();
    });

    test('next with previous index', () => {
        const lua = tstl(
            target,
            `
            const [nextKey, nextValue] = next({ a: "b", c: "d" }, "a");
        `
        );

        expect(lua).toMatchSnapshot();
    });

    test('pairs', () => {
        const lua = tstl(
            target,
            `
            for (const [k, v] of pairs({ foo: "bar", baz: "bur" })) {
                print(k, v);
            }
        `
        );

        expect(lua).toMatchSnapshot();
    });

    test('pairs with LuaTable', () => {
        const lua = tstl(
            target,
            `
            const tbl = new LuaTable<string, string>();
            tbl.set("foo", "bar");
            tbl.set("baz", "bur");
            const takesStr = (str: string) => {};
            for (const [k, v] of pairs(tbl)) {
                takesStr(k);
                takesStr(v);
            }
        `
        );

        expect(lua).toMatchSnapshot();
    });

    test('pcall', () => {
        const lua = tstl(
            target,
            `
            const [success, resultOrMessage] = pcall((a: number) => true, 3);
        `
        );

        expect(lua).toMatchSnapshot();
    });

    test('pcall with context', () => {
        const lua = tstl(
            target,
            `
            const [success, resultOrMessage] = pcall((a: number) => true, {}, 3);
        `
        );

        expect(lua).toMatchSnapshot();
    });

    test('rawget', () => {
        const lua = tstl(
            target,
            `
            const value = rawget({ foo: "bar" }, "foo");
        `
        );

        expect(lua).toMatchSnapshot();
    });

    test('select', () => {
        const lua = tstl(
            target,
            `
            const values = select(2, "a", "b", "c");
        `
        );

        expect(lua).toMatchSnapshot();
    });

    test('select destructured', () => {
        const lua = tstl(
            target,
            `
            const [b, c] = select(2, "a", "b", "c");
        `
        );

        expect(lua).toMatchSnapshot();
    });

    test('select with #', () => {
        const lua = tstl(
            target,
            `
            const count = select("#", "a", "b", "c");
        `
        );

        expect(lua).toMatchSnapshot();
    });

    test('setmetatable with table index', () => {
        const lua = tstl(
            target,
            `
            const tbl = setmetatable({}, {__index: {foo: "bar"}});
            const takesStr = (s: string) => {};
            takesStr(tbl.foo);
            `
        );

        expect(lua).toMatchSnapshot();
    });

    test('setmetatable with function index', () => {
        const lua = tstl(
            target,
            `
            const tbl = setmetatable({}, {__index: (key: string) => key + "bar"});
            const takesStr = (s: string) => {};
            takesStr(tbl.foo);
            `
        );

        expect(lua).toMatchSnapshot();
    });

    test('setmetatable with no index', () => {
        const lua = tstl(
            target,
            `
            const tbl = setmetatable({});
            `
        );

        expect(lua).toMatchSnapshot();
    });

    test('tonumber', () => {
        const lua = tstl(
            target,
            `
            const number = tonumber("213.4");
        `
        );

        expect(lua).toMatchSnapshot();
    });

    test('tonumber with base', () => {
        const lua = tstl(
            target,
            `
            const number = tonumber("213.4", 5);
        `
        );

        expect(lua).toMatchSnapshot();
    });

    test('tostring', () => {
        const lua = tstl(
            target,
            `
            const str = tostring(213.4);
        `
        );

        expect(lua).toMatchSnapshot();
    });

    test('type', () => {
        const lua = tstl(
            target,
            `
            const t = type(213.4);
        `
        );

        expect(lua).toMatchSnapshot();
    });
});
