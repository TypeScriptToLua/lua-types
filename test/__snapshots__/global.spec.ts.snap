// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lua version 5.1 / global assert 1`] = `"assert({bla = \\"not false\\"})"`;

exports[`Lua version 5.1 / global assert 2`] = `"assert(false, \\"assert message\\")"`;

exports[`Lua version 5.1 / global getmetatable 1`] = `
"metatable = getmetatable({})
add = metatable.__add"
`;

exports[`Lua version 5.1 / global ipairs 1`] = `
"for i, v in ipairs({1, 2, 3}) do
    print(i, v)
end"
`;

exports[`Lua version 5.1 / global next 1`] = `"nextKey, nextValue = next({})"`;

exports[`Lua version 5.1 / global next with previous index 1`] = `"nextKey, nextValue = next({a = \\"b\\", c = \\"d\\"}, \\"a\\")"`;

exports[`Lua version 5.1 / global pairs 1`] = `
"for k, v in pairs({foo = \\"bar\\", baz = \\"bur\\"}) do
    print(k, v)
end"
`;

exports[`Lua version 5.1 / global pcall 1`] = `
"success, resultOrMessage = pcall(
    function(a) return true end,
    3
)"
`;

exports[`Lua version 5.1 / global pcall with context 1`] = `
"success, resultOrMessage = pcall(
    function(____, a) return true end,
    {},
    3
)"
`;

exports[`Lua version 5.1 / global rawget 1`] = `"value = rawget({foo = \\"bar\\"}, \\"foo\\")"`;

exports[`Lua version 5.1 / global select 1`] = `
"values = {
    select(2, \\"a\\", \\"b\\", \\"c\\")
}"
`;

exports[`Lua version 5.1 / global select destructured 1`] = `"b, c = select(2, \\"a\\", \\"b\\", \\"c\\")"`;

exports[`Lua version 5.1 / global select with # 1`] = `"count = select(\\"#\\", \\"a\\", \\"b\\", \\"c\\")"`;

exports[`Lua version 5.1 / global tonumber 1`] = `"number = tonumber(\\"213.4\\")"`;

exports[`Lua version 5.1 / global tonumber with base 1`] = `"number = tonumber(\\"213.4\\", 5)"`;

exports[`Lua version 5.1 / global tostring 1`] = `"str = tostring(213.4)"`;

exports[`Lua version 5.1 / global type 1`] = `"t = type(213.4)"`;

exports[`Lua version 5.2 / global assert 1`] = `"assert({bla = \\"not false\\"})"`;

exports[`Lua version 5.2 / global assert 2`] = `"assert(false, \\"assert message\\")"`;

exports[`Lua version 5.2 / global getmetatable 1`] = `
"metatable = getmetatable({})
add = metatable.__add"
`;

exports[`Lua version 5.2 / global ipairs 1`] = `
"for i, v in ipairs({1, 2, 3}) do
    print(i, v)
end"
`;

exports[`Lua version 5.2 / global next 1`] = `"nextKey, nextValue = next({})"`;

exports[`Lua version 5.2 / global next with previous index 1`] = `"nextKey, nextValue = next({a = \\"b\\", c = \\"d\\"}, \\"a\\")"`;

exports[`Lua version 5.2 / global pairs 1`] = `
"for k, v in pairs({foo = \\"bar\\", baz = \\"bur\\"}) do
    print(k, v)
end"
`;

exports[`Lua version 5.2 / global pcall 1`] = `
"success, resultOrMessage = pcall(
    function(a) return true end,
    3
)"
`;

exports[`Lua version 5.2 / global pcall with context 1`] = `
"success, resultOrMessage = pcall(
    function(____, a) return true end,
    {},
    3
)"
`;

exports[`Lua version 5.2 / global rawget 1`] = `"value = rawget({foo = \\"bar\\"}, \\"foo\\")"`;

exports[`Lua version 5.2 / global select 1`] = `
"values = {
    select(2, \\"a\\", \\"b\\", \\"c\\")
}"
`;

exports[`Lua version 5.2 / global select destructured 1`] = `"b, c = select(2, \\"a\\", \\"b\\", \\"c\\")"`;

exports[`Lua version 5.2 / global select with # 1`] = `"count = select(\\"#\\", \\"a\\", \\"b\\", \\"c\\")"`;

exports[`Lua version 5.2 / global tonumber 1`] = `"number = tonumber(\\"213.4\\")"`;

exports[`Lua version 5.2 / global tonumber with base 1`] = `"number = tonumber(\\"213.4\\", 5)"`;

exports[`Lua version 5.2 / global tostring 1`] = `"str = tostring(213.4)"`;

exports[`Lua version 5.2 / global type 1`] = `"t = type(213.4)"`;

exports[`Lua version 5.3 / global assert 1`] = `"assert({bla = \\"not false\\"})"`;

exports[`Lua version 5.3 / global assert 2`] = `"assert(false, \\"assert message\\")"`;

exports[`Lua version 5.3 / global getmetatable 1`] = `
"metatable = getmetatable({})
add = metatable.__add"
`;

exports[`Lua version 5.3 / global ipairs 1`] = `
"for i, v in ipairs({1, 2, 3}) do
    print(i, v)
end"
`;

exports[`Lua version 5.3 / global next 1`] = `"nextKey, nextValue = next({})"`;

exports[`Lua version 5.3 / global next with previous index 1`] = `"nextKey, nextValue = next({a = \\"b\\", c = \\"d\\"}, \\"a\\")"`;

exports[`Lua version 5.3 / global pairs 1`] = `
"for k, v in pairs({foo = \\"bar\\", baz = \\"bur\\"}) do
    print(k, v)
end"
`;

exports[`Lua version 5.3 / global pcall 1`] = `
"success, resultOrMessage = pcall(
    function(a) return true end,
    3
)"
`;

exports[`Lua version 5.3 / global pcall with context 1`] = `
"success, resultOrMessage = pcall(
    function(____, a) return true end,
    {},
    3
)"
`;

exports[`Lua version 5.3 / global rawget 1`] = `"value = rawget({foo = \\"bar\\"}, \\"foo\\")"`;

exports[`Lua version 5.3 / global select 1`] = `
"values = {
    select(2, \\"a\\", \\"b\\", \\"c\\")
}"
`;

exports[`Lua version 5.3 / global select destructured 1`] = `"b, c = select(2, \\"a\\", \\"b\\", \\"c\\")"`;

exports[`Lua version 5.3 / global select with # 1`] = `"count = select(\\"#\\", \\"a\\", \\"b\\", \\"c\\")"`;

exports[`Lua version 5.3 / global tonumber 1`] = `"number = tonumber(\\"213.4\\")"`;

exports[`Lua version 5.3 / global tonumber with base 1`] = `"number = tonumber(\\"213.4\\", 5)"`;

exports[`Lua version 5.3 / global tostring 1`] = `"str = tostring(213.4)"`;

exports[`Lua version 5.3 / global type 1`] = `"t = type(213.4)"`;

exports[`Lua version 5.4 / global assert 1`] = `"assert({bla = \\"not false\\"})"`;

exports[`Lua version 5.4 / global assert 2`] = `"assert(false, \\"assert message\\")"`;

exports[`Lua version 5.4 / global getmetatable 1`] = `
"metatable = getmetatable({})
add = metatable.__add"
`;

exports[`Lua version 5.4 / global ipairs 1`] = `
"for i, v in ipairs({1, 2, 3}) do
    print(i, v)
end"
`;

exports[`Lua version 5.4 / global next 1`] = `"nextKey, nextValue = next({})"`;

exports[`Lua version 5.4 / global next with previous index 1`] = `"nextKey, nextValue = next({a = \\"b\\", c = \\"d\\"}, \\"a\\")"`;

exports[`Lua version 5.4 / global pairs 1`] = `
"for k, v in pairs({foo = \\"bar\\", baz = \\"bur\\"}) do
    print(k, v)
end"
`;

exports[`Lua version 5.4 / global pcall 1`] = `
"success, resultOrMessage = pcall(
    function(a) return true end,
    3
)"
`;

exports[`Lua version 5.4 / global pcall with context 1`] = `
"success, resultOrMessage = pcall(
    function(____, a) return true end,
    {},
    3
)"
`;

exports[`Lua version 5.4 / global rawget 1`] = `"value = rawget({foo = \\"bar\\"}, \\"foo\\")"`;

exports[`Lua version 5.4 / global select 1`] = `
"values = {
    select(2, \\"a\\", \\"b\\", \\"c\\")
}"
`;

exports[`Lua version 5.4 / global select destructured 1`] = `"b, c = select(2, \\"a\\", \\"b\\", \\"c\\")"`;

exports[`Lua version 5.4 / global select with # 1`] = `"count = select(\\"#\\", \\"a\\", \\"b\\", \\"c\\")"`;

exports[`Lua version 5.4 / global tonumber 1`] = `"number = tonumber(\\"213.4\\")"`;

exports[`Lua version 5.4 / global tonumber with base 1`] = `"number = tonumber(\\"213.4\\", 5)"`;

exports[`Lua version 5.4 / global tostring 1`] = `"str = tostring(213.4)"`;

exports[`Lua version 5.4 / global type 1`] = `"t = type(213.4)"`;

exports[`Lua version JIT / global assert 1`] = `"assert({bla = \\"not false\\"})"`;

exports[`Lua version JIT / global assert 2`] = `"assert(false, \\"assert message\\")"`;

exports[`Lua version JIT / global getmetatable 1`] = `
"metatable = getmetatable({})
add = metatable.__add"
`;

exports[`Lua version JIT / global ipairs 1`] = `
"for i, v in ipairs({1, 2, 3}) do
    print(i, v)
end"
`;

exports[`Lua version JIT / global next 1`] = `"nextKey, nextValue = next({})"`;

exports[`Lua version JIT / global next with previous index 1`] = `"nextKey, nextValue = next({a = \\"b\\", c = \\"d\\"}, \\"a\\")"`;

exports[`Lua version JIT / global pairs 1`] = `
"for k, v in pairs({foo = \\"bar\\", baz = \\"bur\\"}) do
    print(k, v)
end"
`;

exports[`Lua version JIT / global pcall 1`] = `
"success, resultOrMessage = pcall(
    function(a) return true end,
    3
)"
`;

exports[`Lua version JIT / global pcall with context 1`] = `
"success, resultOrMessage = pcall(
    function(____, a) return true end,
    {},
    3
)"
`;

exports[`Lua version JIT / global rawget 1`] = `"value = rawget({foo = \\"bar\\"}, \\"foo\\")"`;

exports[`Lua version JIT / global select 1`] = `
"values = {
    select(2, \\"a\\", \\"b\\", \\"c\\")
}"
`;

exports[`Lua version JIT / global select destructured 1`] = `"b, c = select(2, \\"a\\", \\"b\\", \\"c\\")"`;

exports[`Lua version JIT / global select with # 1`] = `"count = select(\\"#\\", \\"a\\", \\"b\\", \\"c\\")"`;

exports[`Lua version JIT / global tonumber 1`] = `"number = tonumber(\\"213.4\\")"`;

exports[`Lua version JIT / global tonumber with base 1`] = `"number = tonumber(\\"213.4\\", 5)"`;

exports[`Lua version JIT / global tostring 1`] = `"str = tostring(213.4)"`;

exports[`Lua version JIT / global type 1`] = `"t = type(213.4)"`;
